sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala

```diff
@@ -0,0 +1,445 @@
+import org.apache.spark.sql.catalyst.util.ArrayData
+import org.apache.spark.sql.execution.streaming.StatefulOperatorStateInfo
+import org.apache.spark.sql.types.{ArrayType, StructType, TimestampType}
+import org.apache.spark.util.NextIterator
+
+/**
+ * Base trait for state manager purposed to be used from streaming session window aggregation.
+ */
+sealed trait StreamingSessionWindowStateManager extends Serializable {
+
+  def getKey(row: InternalRow): UnsafeRow
+
+  def getStartTime(row: InternalRow): Long
+
+  /**
+   * Returns all stored keys.
+   */
+  def getAllKeys(): Iterator[UnsafeRow]
+
+  /**
+   * Returns a list of states for the key. These states are candidates for session window
+   * merging.
+   */
+  def getStates(key: UnsafeRow): Seq[UnsafeRow]
+
+  /**
+   * Returns a list of start times for session windows belonging to the given key.
+   */
+  def getStartTimeList(key: UnsafeRow): Seq[Long]
+
+  /**
+   * Returns the state of given key and start time.
+   */
+  def getState(key: UnsafeRow, startTime: Long): UnsafeRow
+
+  /**
+   * Returns a list of states for all keys.
+   */
+  def getStates(): Seq[UnsafeRow]
+
+  /**
+   * Puts a state row into the state with given key and start time. Note that this method
+   * does not update the start time into the list of start times for the given key. To update
+   * the list, please call `putStartTimeList`.
+   */
+  def putState(key: UnsafeRow, startTime: Long, value: UnsafeRow): Unit
+
+  /**
+   * Puts a list of states for given key. This method will update the list of start times
+   * for the given key.
+   */
+  def putStates(key: UnsafeRow, values: Seq[UnsafeRow]): Unit
+
+  /**
+   * Puts a list of start times of session windows for given key into the state. The list
+   * of start times must be sorted.
+   */
+  def putStartTimeList(key: UnsafeRow, startTimes: Seq[Long]): Unit
+
+  /**
+   * Removes specified state for the given key and start time.
+   */
+  def removeState(key: UnsafeRow, startTime: Long): Unit
+
+  /**
+   * Removes given key from the state store. This method will remove all states associated
+   * with the given key if any.
+   */
+  def removeKey(key: UnsafeRow): Unit
+
+  /**
+   * Given a callback function used to update state store metrics, updates the metrics of all
+   * state stores.
+   */
+  def updateMetrics(updateFunc: StateStoreMetrics => Unit)
+
+  /**
+   * Remove using a predicate on values.
+   *
+   * At a high level, this produces an iterator over the values such that value satisfies the
+   * predicate, where producing an element removes the value from the state store and producing
+   * all elements with a given key updates it accordingly.
+   *
+   * This implies the iterator must be consumed fully without any other operations on this manager
+   * or the underlying store being interleaved.
+   */
+  def removeByValueCondition(removalCondition: UnsafeRow => Boolean): Iterator[UnsafeRow]
```

Comment:
Seems this is not in my implementation. I also don't find it in #31937. Am I missing it?

Reply:
I moved the code on evicting state rows from store physical exec to state store impl, to separate the concern. Streaming join state store also does the same.

I think store exec shouldn't know about the details for state store itself, that said, this may need to be applied to store new state rows as well which is also in store exec. 
(I don't remember I addressed it in this PR. I'm sorry I'm not on my laptop and only mobile for following 3~4 days.)

Reply:
And this change is in #31937.